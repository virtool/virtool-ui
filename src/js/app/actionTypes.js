/**
 * Create a special action type used for requests.
 *
 * The request is an object with three properties assigned with action types used for API requests.
 *
 * - REQUESTED
 * - SUCCEEDED
 * - FAILED
 *
 * @param root {string} the root name of the action type
 * @returns {object} a request-style action type
 */
const createRequestActionType = root => ({
    REQUESTED: `${root}_REQUESTED`,
    SUCCEEDED: `${root}_SUCCEEDED`,
    FAILED: `${root}_FAILED`,
});

// App
export const GET_INITIAL_STATE = createRequestActionType("GET_INITIAL_STATE");

// Account
export const GET_ACCOUNT = createRequestActionType("GET_ACCOUNT");
export const UPDATE_ACCOUNT = createRequestActionType("UPDATE_ACCOUNT");
export const CHANGE_ACCOUNT_PASSWORD = createRequestActionType("CHANGE_ACCOUNT_PASSWORD");
export const LOGIN = createRequestActionType("LOGIN");
export const LOGOUT = createRequestActionType("LOGOUT");
export const RESET_PASSWORD = createRequestActionType("RESET_PASSWORD");

// Administration Settings
export const GET_SETTINGS = createRequestActionType("GET_SETTINGS");
export const UPDATE_SETTINGS = createRequestActionType("UPDATE_SETTINGS");

// Analysis
export const WS_UPDATE_ANALYSIS = "WS_UPDATE_ANALYSIS";
export const SET_ANALYSIS = "SET_ANALYSIS";
export const SET_ANALYSIS_ACTIVE_ID = "SET_ANALYSIS_ACTIVE_ID";
export const SET_SEARCH_IDS = "SET_SEARCH_IDS";
export const SET_AODP_FILTER = "ADD_AODP_FILTER";
export const SET_ANALYSIS_SORT_KEY = "SET_ANALYSIS_SORT_KEY";

// Errors
export const CLEAR_ERROR = "CLEAR_ERROR";

// Files
export const UPLOAD = createRequestActionType("UPLOAD");
export const UPLOAD_SAMPLE_FILE = createRequestActionType("UPLOAD_SAMPLE_FILE");
export const UPLOAD_PROGRESS = "UPLOAD_PROGRESS";
export const UPLOAD_FAILED = "UPLOAD_FAILED";
export const REMOVE_UPLOAD = "REMOVE_UPLOAD";

// Forms
export const SET_PERSISTENT_FORM_STATE = "SET_PERSISTENT_FORM_STATE";
export const DELETE_PERSISTENT_FORM_STATE = "DELETE_PERSISTENT_FORM_STATE";
export const SET_REDUX_FORM_STATE = "SET_REDUX_FORM_STATE";

// Groups
export const CREATE_GROUP = createRequestActionType("CREATE_GROUP");

// HMMs
export const GET_HMM = createRequestActionType("GET_HMM");

// Indexes
export const GET_INDEX = createRequestActionType("GET_INDEX");
export const CREATE_INDEX = createRequestActionType("CREATE_INDEX");

// Jobs
export const GET_JOB = createRequestActionType("GET_JOB");

// OTU
export const GET_OTU = createRequestActionType("GET_OTU");
export const CREATE_OTU = createRequestActionType("CREATE_OTU");
export const EDIT_OTU = createRequestActionType("EDIT_OTU");
export const ADD_ISOLATE = createRequestActionType("ADD_ISOLATE");
export const EDIT_ISOLATE = createRequestActionType("EDIT_ISOLATE");
export const ADD_SEQUENCE = createRequestActionType("ADD_SEQUENCE");
export const EDIT_SEQUENCE = createRequestActionType("EDIT_SEQUENCE");

// Refs
export const GET_REFERENCE = createRequestActionType("GET_REFERENCE");

// Samples
export const GET_SAMPLE = createRequestActionType("GET_SAMPLE");
export const CREATE_SAMPLE = createRequestActionType("CREATE_SAMPLE");
export const UPDATE_SAMPLE = createRequestActionType("UPDATE_SAMPLE");

// Subtraction
export const GET_SUBTRACTION = createRequestActionType("GET_SUBTRACTION");
export const CREATE_SUBTRACTION = createRequestActionType("CREATE_SUBTRACTION");

// Users
export const FIND_USERS = createRequestActionType("FIND_USERS");
export const GET_USER = createRequestActionType("GET_USER");
export const CREATE_USER = createRequestActionType("CREATE_USER");
export const CREATE_FIRST_USER = createRequestActionType("CREATE_FIRST_USER");
export const EDIT_USER = createRequestActionType("EDIT_USER");
