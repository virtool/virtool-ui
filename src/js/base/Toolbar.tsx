import React from "react";
import { cn } from "@utils/utils";

type ToolbarProps = {
    children: React.ReactNode;
    className?: string;
};

export default function Toolbar({ children, className }: ToolbarProps) {
    return <div className={cn(className, "flex gap-2 items-stretch mb-3")}>{children}</div>;
}
