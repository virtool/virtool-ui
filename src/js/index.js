import * as Sentry from "@sentry/browser";
import { createBrowserHistory } from "history";
import "./nonce";
import "normalize.css";
import React from "react";
import ReactDOM from "react-dom";
import App from "./app/App";
import { createAppStore } from "./app/reducer";
import { Request } from "./app/request";
import { setInitialState } from "./app/actions";
if (module.hot) {
    module.hot.accept(err => {
        throw err;
    });
}
const sentryInit = res => {
    const dev = res.body.dev;
    if (!dev) {
        Sentry.init({
            dsn: "https://d9ea493cb0f34ad4a141da5506e6b03b@sentry.io/220541"
        });

        window.Sentry = Sentry;
    }
    return res;
};

window.captureException = error => (window.virtool.dev ? console.error(error) : window.Sentry.captureException(error));
const history = createBrowserHistory();
window.store = createAppStore(history);

Request.get("/api")
    .then(sentryInit)
    .then(({ body }) => window.store.dispatch(setInitialState(body)));

ReactDOM.render(<App store={window.store} history={history} />, document.getElementById("app-container"));
