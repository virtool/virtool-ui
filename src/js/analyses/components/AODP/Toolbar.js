import { toNumber } from "lodash-es";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { Input, SearchInput, Toolbar } from "../../../base";
import { setAnalysisSortKey, setAODPFilter, setSearchIds } from "../../actions";
import { getFuse } from "../../selectors";
import { AnalysisViewerSort } from "../Viewer/Sort";

const StyledAODPToolbar = styled(Toolbar)`
    input[type="number"] {
        max-width: 130px;
    }
`;

export const AODPToolBar = ({ filterAODP, fuse, id, sortKey, onSearch, onSelect, onSetFilter }) => {
    const handleChangeSearch = useCallback(
        e => {
            onSearch(e.target.value, fuse);
        },
        [id]
    );

    const handleChangeFilter = e => {
        onSetFilter(toNumber(e.target.value));
    };

    return (
        <StyledAODPToolbar>
            <SearchInput onChange={handleChangeSearch} onKeyDown={e => e.stopPropagation()} placeholder="Name" />
            <AnalysisViewerSort workflow="aodp" sortKey={sortKey} onSelect={onSelect} />
            <Input
                type="number"
                placeholder="Minimum Identity"
                min={0.5}
                max={1.0}
                step={0.01}
                value={filterAODP.toFixed(2)}
                onChange={e => handleChangeFilter(e)}
            />
        </StyledAODPToolbar>
    );
};

const mapStateToProps = state => ({
    filterAODP: state.analyses.filterAODP,
    fuse: getFuse(state),
    id: state.analyses.detail.id,
    sortKey: state.analyses.sortKey
});

const mapDispatchToProps = dispatch => ({
    onSearch: (term, fuse) => {
        dispatch(setSearchIds(term ? fuse.search(term) : null));
    },

    onSelect: sortKey => {
        dispatch(setAnalysisSortKey(sortKey));
    },

    onSetFilter: identity => {
        dispatch(setAODPFilter(identity));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(AODPToolBar);
