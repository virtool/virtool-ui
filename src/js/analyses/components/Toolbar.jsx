import React from "react";
import { connect } from "react-redux";
import { pushState } from "../../app/actions";
import { Button, InputSearch, Toolbar } from "../../base";
import { getCanModify } from "../../samples/selectors";
import { findAnalyses } from "../actions";

export function AnalysesToolbar({ canModify, onFind, onShowCreate, page, sampleId, term }) {
    return (
        <Toolbar>
            <InputSearch value={term} onChange={e => onFind(sampleId, e.target.value, page)} />
            <Button
                icon="plus-square fa-fw"
                tip="New Analysis"
                color="blue"
                onClick={() => onShowCreate(sampleId)}
                disabled={!canModify}
            />
        </Toolbar>
    );
}

export function mapStateToProps(state) {
    return {
        canModify: getCanModify(state),
        page: state.analyses.page,
        sampleId: state.samples.detail.id,
        term: state.analyses.term,
    };
}

export function mapDispatchToProps(dispatch) {
    return {
        onFind: (sampleId, term, page) => {
            dispatch(findAnalyses(sampleId, term, page));
        },
        onShowCreate: sampleId => {
            dispatch(pushState({ createAnalysis: sampleId }));
        },
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(AnalysesToolbar);
