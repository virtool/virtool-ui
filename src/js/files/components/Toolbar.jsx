import { forEach } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { checkAdminRoleOrPermission } from "../../administration/utils";
import { Alert, Icon, UploadBar } from "../../base";
import { createRandomString } from "../../utils/utils";
import { upload } from "../actions";

export const UploadToolbar = ({ canUpload, onDrop, fileType, message, validationRegex, tip }) => {
    if (!canUpload) {
        return (
            <Alert color="orange" level>
                <Icon name="exclamation-circle" />
                <span>
                    <strong>You do not have permission to upload files.</strong>
                    <span> Contact an administrator.</span>
                </span>
            </Alert>
        );
    }

    const handleDrop = acceptedFiles => {
        onDrop(fileType, acceptedFiles);
    };

    const validateExtensions = file => (validationRegex.test(file.name) ? null : { code: "Invalid file type" });

    return (
        <UploadBar
            onDrop={handleDrop}
            message={message || "Drag file here to upload."}
            validator={validationRegex ? validateExtensions : null}
            tip={tip}
        />
    );
};

export const mapStateToProps = state => ({
    canUpload: checkAdminRoleOrPermission(state, "upload_file")
});

export const mapDispatchToProps = dispatch => ({
    onDrop: (fileType, acceptedFiles) => {
        forEach(acceptedFiles, file => {
            const localId = createRandomString();
            dispatch(upload(localId, file, fileType));
        });
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(UploadToolbar);
