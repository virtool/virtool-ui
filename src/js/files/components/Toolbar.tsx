import { forEach } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { useCheckAdminRoleOrPermission } from "../../administration/hooks";
import { Alert, Icon, UploadBar } from "../../base";
import { Permission } from "../../groups/types";
import { createRandomString } from "../../utils/utils";
import { upload } from "../actions";

type UploadToolbarProps = {
    onDrop: (fileType: string, acceptedFiles: File[]) => void;
    fileType: string;
    tip?: string;
    message?: React.ReactNode;
    /* For validating file type */
    validationRegex?: RegExp;
};

/*
 * Renders an UploadBar if the user has permission to upload files.
 */
export function UploadToolbar({ onDrop, fileType, message, validationRegex }: UploadToolbarProps) {
    const { hasPermission: canUpload } = useCheckAdminRoleOrPermission(Permission.upload_file);

    if (!canUpload) {
        return (
            <Alert color="orange" level>
                <Icon name="exclamation-circle" />
                <span>
                    <strong>You do not have permission to upload files.</strong>
                    <span> Contact an administrator.</span>
                </span>
            </Alert>
        );
    }

    function handleDrop(acceptedFiles: File[]) {
        onDrop(fileType, acceptedFiles);
    }

    function validateExtensions(file: File) {
        return validationRegex.test(file.name) ? null : { code: "Invalid file type", message: "File type not allowed" };
    }

    return (
        <UploadBar
            onDrop={handleDrop}
            message={message || "Drag file here to upload."}
            validator={validationRegex ? validateExtensions : null}
        />
    );
}

export function mapDispatchToProps(dispatch) {
    return {
        onDrop: (fileType: string, acceptedFiles: File[]) => {
            forEach(acceptedFiles, (file: File) => {
                const localId = createRandomString();
                dispatch(upload(localId, file, fileType));
            });
        },
    };
}

export default connect(null, mapDispatchToProps)(UploadToolbar);
