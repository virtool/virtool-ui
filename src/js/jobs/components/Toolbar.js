import React from "react";
import { connect } from "react-redux";
import { SearchInput, Toolbar } from "../../base";

import { updateJobsSearch } from "../actions";
import { getTermFromURL } from "../selectors";

export const JobsToolbar = ({ onUpdateJobsSearch, term }) => (
    <Toolbar>
        <SearchInput value={term} onChange={onUpdateJobsSearch} placeholder="User or workflow" />
    </Toolbar>
);

export const mapStateToProps = state => ({
    term: getTermFromURL(state)
});

export const mapDispatchToProps = dispatch => ({
    onUpdateJobsSearch: e => {
        dispatch(updateJobsSearch({ find: e.target.value }));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(JobsToolbar);
