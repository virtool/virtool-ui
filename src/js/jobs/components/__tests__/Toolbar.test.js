import React from "react";
import { UPDATE_JOB_SEARCH } from "../../../app/actionTypes";
import { JobsToolbar, mapDispatchToProps, mapStateToProps } from "../Toolbar.js";

jest.mock("../../../utils/utils");

describe("<JobsToolbar />", () => {
    let props;

    beforeEach(() => {
        props = {
            onFind: jest.fn(),
            term: "foo"
        };
    });

    it("should render", () => {
        const wrapper = shallow(<JobsToolbar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("mapStateToProps", () => {
    it("should return correct props", () => {
        const state = {
            jobs: {
                term: "bar"
            },
            account: {
                administrator: true,
                permissions: true
            },
            router: { location: new window.URL("https://www.virtool.ca?find=bar") }
        };
        const props = mapStateToProps(state);
        expect(props).toEqual({
            term: "bar"
        });
    });
});

describe("mapDispatchToProps", () => {
    it("should return onFind() in props", () => {
        const dispatch = jest.fn();
        const props = mapDispatchToProps(dispatch);
        const e = {
            target: {
                value: "Foo"
            }
        };
        props.onUpdateJobsSearch(e, "foo", "bar");
        expect(dispatch).toHaveBeenCalledWith({
            type: UPDATE_JOB_SEARCH,
            payload: { searchParameters: { find: "Foo" } }
        });
    });
});
