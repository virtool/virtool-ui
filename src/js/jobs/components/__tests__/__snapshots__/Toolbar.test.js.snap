// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JobsToolbar /> should render 1`] = `
<Toolbar>
  <SearchInput
    placeholder="User or workflow"
    value="foo"
  />
</Toolbar>
`;
